x/valset-pref/msg_server_test.go

```diff
@@ -0,0 +1,110 @@
+package keeper_test
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	valPref "github.com/osmosis-labs/osmosis/v12/x/valset-pref"
+	"github.com/osmosis-labs/osmosis/v12/x/valset-pref/types"
+)
+
+func (suite *KeeperTestSuite) TestSetValidatorSetPreference() {
+	suite.SetupTest()
+
+	// setup 3 validators
+	valAddrs := suite.SetupMultipleValidators(3)
+
+	tests := []struct {
+		name        string
+		delegator   sdk.AccAddress
+		preferences []types.ValidatorPreference
+		expectPass  bool
+	}{
+		{
+			name:      "creation of new validator set",
+			delegator: sdk.AccAddress([]byte("addr1---------------")),
+			preferences: []types.ValidatorPreference{
+				{
+					ValOperAddress: valAddrs[0],
+					Weight:         sdk.NewDecWithPrec(5, 1),
+				},
+				{
+					ValOperAddress: valAddrs[1],
+					Weight:         sdk.NewDecWithPrec(3, 1),
+				},
+				{
+					ValOperAddress: valAddrs[2],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name:      "update existing validator with same valAddr and weights",
+			delegator: sdk.AccAddress([]byte("addr1---------------")),
+			preferences: []types.ValidatorPreference{
+				{
+					ValOperAddress: valAddrs[1],
+					Weight:         sdk.NewDecWithPrec(3, 1),
+				},
+				{
+					ValOperAddress: valAddrs[0],
+					Weight:         sdk.NewDecWithPrec(5, 1),
+				},
+				{
+					ValOperAddress: valAddrs[2],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+			},
+			expectPass: false,
+		},
+		{
+			name:      "update existing validator with same valAddr but different weights",
+			delegator: sdk.AccAddress([]byte("addr1---------------")),
+			preferences: []types.ValidatorPreference{
+				{
+					ValOperAddress: valAddrs[0],
+					Weight:         sdk.NewDecWithPrec(3, 1),
+				},
+				{
+					ValOperAddress: valAddrs[1],
+					Weight:         sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					ValOperAddress: valAddrs[2],
+					Weight:         sdk.NewDecWithPrec(5, 1),
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name:      "create validator set with unknown validator address",
+			delegator: sdk.AccAddress([]byte("addr2---------------")),
+			preferences: []types.ValidatorPreference{
+				{
+					ValOperAddress: "addr1---------------",
+					Weight:         sdk.NewDec(1),
+				},
+			},
+			expectPass: false,
```

Comment:
Should we test a case with valid validators and only one of the ones in the set is invalid?

Reply:
i've done all the validity checks in `ValidateBasic`, so didn't wanted to repeat the same tests 
https://github.com/osmosis-labs/osmosis/pull/2892/files#diff-3f8c2f857e1dbb8c5c670c48e165fea76702e1961b545a96657caecbdbf14504R28